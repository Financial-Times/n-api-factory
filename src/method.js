import { trimObject } from '@financial-times/n-utils';
import { fetch, parseFetchError } from '@financial-times/n-error';

export const defaultHeaders = ({
	API_KEY,
	transactionId,
	requestId,
	systemId = process.env.SYSTEM_CODE,
	user = 'customer',
}) =>
	trimObject({
		'Content-Type': 'application/json',
		'x-api-key': API_KEY,
		'FT-Transaction-Id': transactionId || requestId,
		'X-Origin-System-Id': `https://cmdb.ft.com/systems/${systemId}`,
		'X-Origin-User': user,
	});

export const fetchWithErrorParser = async (url, options) => {
	try {
		const response = await fetch(url, options);

		if (!response.ok) {
			throw response;
		}
		// in case of 'NO CONTENT'
		const contentLength = parseInt(response.headers.get('content-length'), 10);
		if (response.status === 204 || contentLength === 0) {
			return undefined;
		}

		const contentType = response.headers.get('content-type');
		const parseMethod =
			contentType && contentType.includes('application/json') ? 'json' : 'text';
		return await response[parseMethod]();
	} catch (e) {
		const parsed = await parseFetchError(e);
		throw parsed;
	}
};
