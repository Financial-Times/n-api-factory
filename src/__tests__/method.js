import nock from 'nock';

import { fetchWithErrorParser } from '../method';

const config = {
	API_HOST: 'http://mock.com',
	API_KEY: 'mock.key',
};

describe('fetchWithErrorParser', () => {
	afterEach(() => {
		nock.cleanAll();
		jest.clearAllMocks();
	});

	describe('return the parsed data correctly if response.ok for ', () => {
		it('json', async () => {
			const res = { foo: 'bar' };
			nock(config.API_HOST)
				.get('/')
				.reply(200, res);
			const data = await fetchWithErrorParser(`${config.API_HOST}`);
			expect(data).toEqual(res);
		});

		it('text', async () => {
			const res = 'some message';
			nock(config.API_HOST)
				.get('/')
				.reply(200, res);
			const data = await fetchWithErrorParser(`${config.API_HOST}`);
			expect(data).toEqual(res);
		});
	});

	describe('return undefined if', () => {
		it('response.status 204', async () => {
			nock(config.API_HOST)
				.get('/')
				.reply(204);

			const data = await fetchWithErrorParser(`${config.API_HOST}`);
			expect(data).toBe(undefined);
		});

		it('response.body is emtpy', async () => {
			nock(config.API_HOST)
				.get('/')
				.reply(200, '', { 'Content-Length': '0' });

			const data = await fetchWithErrorParser(`${config.API_HOST}`);
			expect(data).toBe(undefined);
		});
	});

	it('return formatted error correctly if response is error', async () => {
		nock(config.API_HOST)
			.get('/')
			.reply(404, { message: 'some message' });
		try {
			await fetchWithErrorParser(`${config.API_HOST}`);
			throw new Error('it should throw an exception');
		} catch (e) {
			expect(e).toEqual({
				category: 'FETCH_RESPONSE_ERROR',
				contentType: 'application/json',
				status: 404,
				message: {
					message: 'some message',
				},
			});
		}
	});
});
