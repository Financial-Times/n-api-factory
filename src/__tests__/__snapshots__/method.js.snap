// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchWithErrorParser return formatted error correctly if response is error 1`] = `
NError {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "application/json",
  "message": Object {
    "message": "some message",
  },
  "status": 404,
}
`;
